### PR TITLE
Allow overloading properties with non-overlapping availability

### PR DESCRIPTION
**Pull request by slavapestov**
This generalizes the logic in https://github.com/apple/swift/pull/8009 which was recently added by @jrose-apple to allow "overloading" properties with different types as long as they have non-overlapping availability.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: test/SILGen/availability_overloads.swift [hunk 1]
```diff
@@ -1,25 +1,44 @@
...
 
 // This is a "don't crash with duplicate definition errors" test.
 // We care about being able to express each of these "redeclarations" when the
 // availability doesn't overlap.
 
-class BeforeAndAfter {
+import availability_overloads_other
+
+// FIXME: Should this work?
+#if false
```
Comment (slavapestov) [inline test/SILGen/availability_overloads.swift]: @jrose-apple I moved the BeforeAndAfter class into a separate module to make sure we can serialize this stuff properly, too.

Do you think we should be able to override these properties? I guess override checking needs a 'non-overlapping availability' check of some kind too.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: lib/Sema/TypeCheckDecl.cpp [hunk 2]
```diff
@@ -975,6 +976,19 @@ static void checkRedeclaration(TypeChecker &tc, ValueDecl *current) {
           isAcceptableVersionBasedChange = true;
         }
       }
+      // - or they are computed properties of different types,
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckDecl.cpp]: I think *one* of them can be a stored property, right?

File: lib/Sema/TypeCheckDecl.cpp [hunk 3]
```diff
@@ -975,6 +976,19 @@ static void checkRedeclaration(TypeChecker &tc, ValueDecl *current) {
...
+      {
+        const auto *currentVD = dyn_cast<VarDecl>(current);
+        const auto *otherVD = dyn_cast<VarDecl>(other);
+        if (currentVD && otherVD &&
+            !currentVD->hasStorage() &&
+            !otherVD->hasStorage() &&
+            !currentVD->getInterfaceType()->isEqual(
+              otherVD->getInterfaceType())) {
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckDecl.cpp]: This seems like it might be incorrect if the "redeclaration" is across extension boundaries.

**Review by jrose-apple — COMMENTED** [review]
That was quick! Do we need to change mangling as well, though? Properties don't have symbols but they do have USRs.

File: lib/Sema/TypeCheckDecl.cpp (hunk 3 above)
Comment (slavapestov) [inline lib/Sema/TypeCheckDecl.cpp]: You mean with generic classes and extensions? Do you have an example?

File: lib/Sema/TypeCheckDecl.cpp (hunk 2 above)
Comment (slavapestov) [inline lib/Sema/TypeCheckDecl.cpp]: Yes, but what about DI? Even the unavailable property has to be initialized.

**Comment by slavapestov** [comment]
Yeah I think the right thing to do is to always mangle the type with the property. I *think* the only time it would come up is with USRs and stored property field offset symbols (but those are not public AFAIK). I'll make a follow up patch. Thanks for the review!

File: lib/Sema/TypeCheckDecl.cpp (hunk 2 above)
Comment (jrose-apple) [inline lib/Sema/TypeCheckDecl.cpp]: That's fine as long as the one that's available is the one in the current language mode.

File: lib/Sema/TypeCheckDecl.cpp (hunk 3 above)
Comment (jrose-apple) [inline lib/Sema/TypeCheckDecl.cpp]: Something like this, maybe?

```swift
class A<T, U> {}

extension A {
  @available(swift 4)
  var x: T? { return nil }
}

extension A where T == U {
  @available(swift, obsoleted: 4)
  var x: U? { return nil }
}
```

I'm just making things up; I don't actually have a test case. It just set off warning bells.

**Pull request merged**